crates/api/src/config.rs

```diff
@@ -84,100 +253,91 @@ mod test {
+        c.set_module_config(&serde_json::json!({"bananas": 42}))
+            .unwrap();
+    }
+
+    #[test]
+    fn warns_no_runtime_cb() {
+        // this test will never fail,
+        // but we can check it traces correctly manually
+
+        kitsune2_test_utils::enable_tracing();
 
-        impl ModConfig for Mod1 {}
+        let c = Config::default();
+        c.set_module_config(&serde_json::json!({"apples": "red"}))
+            .unwrap();
+        c.mark_runtime();
+        c.set_module_config(&serde_json::json!({"apples": "green"}))
+            .unwrap();
+        c.set_module_config(&serde_json::json!({"bananas": 42}))
+            .unwrap();
+    }
 
+    #[test]
+    fn config_usage_example() {
         #[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
-        #[serde(default)]
-        struct Mod2 {
-            #[serde(rename = "#p_c", skip_deserializing)]
-            doc_p_a: &'static str,
-            #[serde(default)]
-            p_c: u32,
-            #[serde(default)]
-            p_d: String,
+        #[serde(rename_all = "camelCase")]
+        struct SubConfig {
+            pub apples: String,
+            pub bananas: u32,
         }
 
-        impl Default for Mod2 {
-            fn default() -> Self {
-                Self {
-                    doc_p_a: "This is a potential pattern for how to document config props.",
-                    p_c: 0,
-                    p_d: "".into(),
-                }
-            }
+        #[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+        #[serde(rename_all = "camelCase")]
+        struct ModConfig {
+            pub my_module: SubConfig,
         }
 
-        impl ModConfig for Mod2 {}
+        let c = Config::default();
 
-        let mut config = Config::default();
-        config
-            .add_default_module_config::<Mod1>("mod1".into())
-            .unwrap();
-        config
-            .add_default_module_config::<Mod2>("mod2".into())
-            .unwrap();
+        let expect = ModConfig {
+            my_module: SubConfig {
+                apples: "red".to_string(),
+                bananas: 42,
+            },
+        };
+
+        c.set_module_config(&expect).unwrap();
 
-        // output the "default" config
-        assert_eq!(
-            r##"{
-  "mod1": {
-    "p_a": 0,
-    "p_b": ""
-  },
-  "mod2": {
-    "#p_c": "This is a potential pattern for how to document config props.",
-    "p_c": 0,
-    "p_d": ""
-  }
-}"##,
-            serde_json::to_string_pretty(&config).unwrap()
-        );
-
-        // ensure we can load a weird config from disk
-        let config: Config = serde_json::from_str(
-            r#"{
-          "modBAD": { "foo": "bar" },
-          "mod1": { "p_b": "test-p_b" },
-          "mod2": { "p_c": 42, "p_d": "test-p_d", "extra": "foo" }
-        }"#,
+        println!("{}", serde_json::to_string_pretty(&c).unwrap());
+
+        let resp: ModConfig = c.get_module_config().unwrap();
```

Comment:
Not quite sure I'm following this part. This is actually transcoding from the internal format to the `ModConfig` struct here?

If so, I wonder if we want to try to be a bit more efficient than calling `get_module_config` always in modules that can have their config change. We always want to use the latest config but don't want to do the deserialize every time we need config.

Reply:
The intention is in general to use `get_module_config` the normal intuitive way, most modules will just get it at the beginning and never bother with updates.

If your module has a config that wants to change at runtime, it should register an update callback which will receive the value directly.

Reply:
Cool okay, so get the config once and put it somewhere then your update functions can update your inner config. Great, happy with that